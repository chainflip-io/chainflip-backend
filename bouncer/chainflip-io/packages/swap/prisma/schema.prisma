// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  channelId             BigInt
  srcChain              Chain
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@unique([issuedBlock, srcChain, channelId])
  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal             @db.Decimal(30, 0)
  intermediateAmount        Decimal?            @db.Decimal(30, 0)
  egressAmount              Decimal?            @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id                  BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId            BigInt
  chain               Chain
  amount              Decimal    @db.Decimal(30, 0)
  scheduledAt         DateTime
  scheduledBlockIndex String
  swap                Swap?      @relation(fields: [swapId], references: [id])
  swapId              BigInt?    @unique
  broadcast           Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId         BigInt?
  createdAt           DateTime   @default(now())
  updatedAt           DateTime   @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id                  BigInt    @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId            BigInt
  chain               Chain
  egresses            Egress[]
  requestedAt         DateTime
  requestedBlockIndex String
  succeededAt         DateTime?
  succeededBlockIndex String?
  abortedAt           DateTime?
  abortedBlockIndex   String?
  createdAt           DateTime  @default(now())
  updatedAt           DateTime  @default(now()) @updatedAt

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
