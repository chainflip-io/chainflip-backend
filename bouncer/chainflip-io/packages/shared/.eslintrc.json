{
  "extends": "../../.eslintrc.json",
  "plugins": ["eslint-plugin-n"],
  "rules": {
    "n/no-process-env": ["error"]
  },
  "overrides": [
    {
      "files": ["*.test.ts", "*.mjs", "*.js", "*.cjs"],
      "rules": {
        "n/no-process-env": ["off"]
      }
    }
  ]
}
