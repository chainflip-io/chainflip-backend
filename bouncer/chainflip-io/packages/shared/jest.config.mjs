// eslint-disable-next-line import/no-relative-packages
import baseConfig from '../../jest.config.mjs';

process.env.RPC_BROKER_HTTPS_URL = 'wss://rpc-broker.example.com';

export default {
  ...baseConfig,
};
