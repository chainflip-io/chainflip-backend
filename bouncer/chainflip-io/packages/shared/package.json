{
  "name": "@chainflip-io/sdk-shared",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "rm -rf dist",
    "build": "tsc -p tsconfig.cjs.json && tsc -p tsconfig.mjs.json",
    "build:abis": "typechain --target=ethers-v5 --out-dir=./src/abis ./abis/**",
    "lint": "pnpm eslint --max-warnings 0 './**/*.ts'",
    "test": "jest",
    "test:ci": "CI=1 pnpm test -- --forceExit"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "peerDependencies": {
    "zod": "^3.21.4"
  },
  "dependencies": {
    "@polkadot/util": "^12.0.1",
    "@polkadot/util-crypto": "^12.0.1",
    "ethers": "5.7.2",
    "rxjs": "^7.8.0",
    "ws": "^8.13.0"
  },
  "devDependencies": {
    "@ethersproject/abi": "^5.7.0",
    "@ethersproject/providers": "^5.7.2",
    "@typechain/ethers-v5": "^11.0.0",
    "@types/ws": "^8.5.4",
    "typechain": "^8.2.0",
    "winston": "^3.8.2",
    "zod": "^3.21.4"
  }
}
