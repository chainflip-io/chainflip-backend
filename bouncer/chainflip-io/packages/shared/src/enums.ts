type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot', 'Arbitrum']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC', 'ARBETH', 'ARBUSDC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'sisyphos',
  'perseverance',
  'mainnet',
  'partnernet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

export const assetChains = {
  [Assets.ETH]: Chains.Ethereum,
  [Assets.FLIP]: Chains.Ethereum,
  [Assets.USDC]: Chains.Ethereum,
  [Assets.BTC]: Chains.Bitcoin,
  [Assets.DOT]: Chains.Polkadot,
  [Assets.ARBETH]: Chains.Arbitrum,
  [Assets.ARBUSDC]: Chains.Arbitrum,
} satisfies Record<Asset, Chain>;

export const assetDecimals = {
  [Assets.DOT]: 10,
  [Assets.ETH]: 18,
  [Assets.FLIP]: 18,
  [Assets.USDC]: 6,
  [Assets.BTC]: 8,
  [Assets.ARBETH]: 18,
  [Assets.ARBUSDC]: 6,
} satisfies Record<Asset, number>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetContractIds: Record<Asset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  [Assets.ETH]: 1,
  [Assets.FLIP]: 2,
  [Assets.USDC]: 3,
  [Assets.DOT]: 4,
  [Assets.BTC]: 5,
  [Assets.ARBETH]: 6,
  [Assets.ARBUSDC]: 7,
};

export const chainAssets = {
  [Chains.Ethereum]: [Assets.ETH, Assets.USDC, Assets.FLIP],
  [Chains.Bitcoin]: [Assets.BTC],
  [Chains.Polkadot]: [Assets.DOT],
  [Chains.Arbitrum]: [Assets.ARBETH, Assets.ARBUSDC],
} satisfies Record<Chain, Asset[]>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainContractIds: Record<Chain, number> = {
  [Chains.Ethereum]: 1,
  [Chains.Polkadot]: 2,
  [Chains.Bitcoin]: 3,
  [Chains.Arbitrum]: 4,
};
