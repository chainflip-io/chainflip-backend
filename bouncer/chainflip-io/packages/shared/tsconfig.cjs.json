{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "CommonJS",
    "outDir": "./dist/cjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
