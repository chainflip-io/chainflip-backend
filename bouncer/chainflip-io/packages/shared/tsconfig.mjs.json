{
  "extends": "./tsconfig",
  "compilerOptions": {
    "module": "ESNext",
    "outDir": "./dist/mjs" /* Redirect output structure to the directory. */,
    "noEmit": false
  }
}
