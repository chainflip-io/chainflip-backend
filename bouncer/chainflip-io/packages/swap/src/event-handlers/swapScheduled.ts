// Set the column in the DB to the block timestamp and the deposit amount.
import assert from 'assert';
import { z } from 'zod';
import { chainflipAssetEnum, u128, u64 } from '@/shared/parsers';
import logger from '../utils/logger';
import { encodedAddress } from './common';
import type { EventHandlerArgs } from '.';

const depositChannelSwapOrigin = z.object({
  __kind: z.literal('DepositChannel'),
  channelId: u64,
  depositAddress: encodedAddress,
});
const vaultSwapOrigin = z.object({
  __kind: z.literal('Vault'),
  txHash: z.string(),
});

const swapScheduledArgs = z.object({
  swapId: u64,
  sourceAsset: chainflipAssetEnum,
  depositAmount: u128,
  destinationAsset: chainflipAssetEnum,
  destinationAddress: encodedAddress,
  origin: z.union([depositChannelSwapOrigin, vaultSwapOrigin]),
});

export type SwapScheduledEvent = z.input<typeof swapScheduledArgs>;

export default async function swapScheduled({
  prisma,
  block,
  event,
}: EventHandlerArgs): Promise<void> {
  const {
    swapId,
    sourceAsset,
    depositAmount,
    destinationAsset,
    destinationAddress,
    origin,
  } = swapScheduledArgs.parse(event.args);

  const newSwapData = {
    depositReceivedBlockIndex: `${block.height}-${event.indexInBlock}`,
    depositAmount: depositAmount.toString(),
    nativeId: swapId,
    depositReceivedAt: new Date(block.timestamp),
  };

  if (origin.__kind === 'DepositChannel') {
    const depositAddress = origin.depositAddress.address;

    const channels = await prisma.swapDepositChannel.findMany({
      where: {
        srcAsset: sourceAsset,
        depositAddress,
        expiryBlock: { gte: block.height },
        issuedBlock: { lte: block.height },
      },
    });
    if (channels.length === 0) {
      logger.info(
        `SwapScheduled: SwapDepositChannel not found for depositAddress ${depositAddress}`,
      );
      return;
    }
    assert(
      channels.length === 1,
      `SwapScheduled: too many active swap intents found for depositAddress ${depositAddress}`,
    );

    const [{ srcAsset, destAddress, destAsset, id }] = channels;

    await prisma.swap.create({
      data: {
        swapDepositChannelId: id,
        srcAsset,
        destAsset,
        destAddress,
        ...newSwapData,
      },
    });
  } else if (origin.__kind === 'Vault') {
    await prisma.swap.create({
      data: {
        srcAsset: sourceAsset,
        destAsset: destinationAsset,
        destAddress: destinationAddress.address,
        txHash: origin.txHash,
        ...newSwapData,
      },
    });
  }
}
