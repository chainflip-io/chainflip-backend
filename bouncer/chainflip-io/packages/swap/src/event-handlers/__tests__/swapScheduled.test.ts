import { Assets } from '@/shared/enums';
import prisma, { SwapDepositChannel } from '../../client';
import swapScheduled from '../swapScheduled';
import {
  createDepositChannel,
  swapScheduledBtcDepositChannelMock,
  swapScheduledDotDepositChannelMock,
  swapScheduledVaultMock,
} from './utils';

describe(swapScheduled, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  describe('deposit channel origin', () => {
    let dotSwapDepositChannel: SwapDepositChannel;
    let btcSwapDepositChannel: SwapDepositChannel;

    beforeEach(async () => {
      dotSwapDepositChannel = await createDepositChannel({
        srcChain: 'Polkadot',
        srcAsset: Assets.DOT,
        destAsset: Assets.BTC,
        depositAddress: '5CGLqaFMheyVcsXz6QEtjtSAi6RcXFaEDJKvovgCdPiZhw11',
        destAddress: 'bcrt1pzjdpc799qa5f7m65hpr66880res5ac3lr6y2chc4jsa',
      });
      btcSwapDepositChannel = await createDepositChannel({
        srcChain: 'Bitcoin',
        srcAsset: Assets.BTC,
        destAsset: Assets.ETH,
        depositAddress: 'bcrt1pzjdpc799qa5f7m65hpr66880res5ac3lr6y2chc4jsa',
        destAddress: '0x41ad2bc63a2059f9b623533d87fe99887d794847',
      });
    });

    it('stores a new swap from a dot deposit channel', async () => {
      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDotDepositChannelMock.block as any,
          event: swapScheduledDotDepositChannelMock.eventContext.event as any,
        });
      });

      const swap = await prisma.swap.findFirstOrThrow({
        where: { swapDepositChannelId: dotSwapDepositChannel.id },
      });

      expect(swap.depositAmount.toString()).toEqual(
        swapScheduledDotDepositChannelMock.eventContext.event.args
          .depositAmount,
      );
      expect(swap).toMatchSnapshot({
        id: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        swapDepositChannelId: expect.any(BigInt),
      });
    });

    it('stores a new swap from a btc deposit channel', async () => {
      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledBtcDepositChannelMock.block as any,
          event: swapScheduledBtcDepositChannelMock.eventContext.event as any,
        });
      });

      const swap = await prisma.swap.findFirstOrThrow({
        where: { swapDepositChannelId: btcSwapDepositChannel.id },
      });

      expect(swap.depositAmount.toString()).toEqual(
        swapScheduledBtcDepositChannelMock.eventContext.event.args
          .depositAmount,
      );
      expect(swap).toMatchSnapshot({
        id: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        swapDepositChannelId: expect.any(BigInt),
      });
    });

    it('does not store a new swap if the deposit channel is expired', async () => {
      await prisma.swapDepositChannel.update({
        where: { id: dotSwapDepositChannel.id },
        data: { expiryBlock: -1 },
      });

      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDotDepositChannelMock.block,
          event: swapScheduledDotDepositChannelMock.eventContext.event as any,
        });
      });

      expect(await prisma.swap.findFirst()).toBeNull();
    });

    it('does not store a new swap if the deposit channel is not found', async () => {
      await prisma.swapDepositChannel.update({
        where: { id: dotSwapDepositChannel.id },
        data: { depositAddress: '0x0' },
      });

      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledDotDepositChannelMock.block,
          event: swapScheduledDotDepositChannelMock.eventContext.event as any,
        });
      });

      expect(await prisma.swap.findFirst()).toBeNull();
    });

    it('does not store a new swap if the deposit channel is not unique', async () => {
      const { id, ...rest } = dotSwapDepositChannel;
      await prisma.swapDepositChannel.create({
        data: { ...rest, channelId: 2n },
      });

      await expect(
        prisma.$transaction(async (client) => {
          await swapScheduled({
            prisma: client,
            block: swapScheduledDotDepositChannelMock.block,
            event: swapScheduledDotDepositChannelMock.eventContext.event as any,
          });
        }),
      ).rejects.toThrowError();

      expect(await prisma.swap.findFirst()).toBeNull();
    });
  });

  describe('smart contract origin', () => {
    it('stores a new swap from a contract deposit', async () => {
      // create a swap after receiving the event
      await prisma.$transaction(async (client) => {
        await swapScheduled({
          prisma: client,
          block: swapScheduledVaultMock.block as any,
          event: swapScheduledVaultMock.eventContext.event as any,
        });
      });

      const swap = await prisma.swap.findFirstOrThrow();

      expect(swap.depositAmount.toString()).toEqual(
        swapScheduledVaultMock.eventContext.event.args.depositAmount,
      );
      expect(swap).toMatchSnapshot({
        id: expect.any(BigInt),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });
    });
  });
});
