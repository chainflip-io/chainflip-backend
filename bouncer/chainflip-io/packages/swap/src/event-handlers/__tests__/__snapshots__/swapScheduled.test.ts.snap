// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swapScheduled deposit channel origin stores a new swap from a btc deposit channel 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "75000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "0x41ad2bc63a2059f9b623533d87fe99887d794847",
  "destAsset": "ETH",
  "egressAmount": null,
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 3n,
  "srcAsset": "BTC",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "txHash": null,
  "updatedAt": Any<Date>,
}
`;

exports[`swapScheduled deposit channel origin stores a new swap from a dot deposit channel 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "125000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "bcrt1pzjdpc799qa5f7m65hpr66880res5ac3lr6y2chc4jsa",
  "destAsset": "BTC",
  "egressAmount": null,
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 1n,
  "srcAsset": "DOT",
  "swapDepositChannelId": Any<BigInt>,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "txHash": null,
  "updatedAt": Any<Date>,
}
`;

exports[`swapScheduled smart contract origin stores a new swap from a contract deposit 1`] = `
{
  "createdAt": Any<Date>,
  "depositAmount": "175000000000000000",
  "depositReceivedAt": 2022-12-06T14:31:33.000Z,
  "depositReceivedBlockIndex": "100-0",
  "destAddress": "5D34dL5prEUaGNQtPPZ3yN5Y6BnkfXunKXXz6fo7ZJbLwRRH",
  "destAsset": "DOT",
  "egressAmount": null,
  "id": Any<BigInt>,
  "intermediateAmount": null,
  "nativeId": 2n,
  "srcAsset": "ETH",
  "swapDepositChannelId": null,
  "swapExecutedAt": null,
  "swapExecutedBlockIndex": null,
  "txHash": "0x1103ebed92b02a278b54789bfabc056e69ad5c6558049364ea23ec2f3bfa0fd9",
  "updatedAt": Any<Date>,
}
`;
